### PR TITLE
[SelectionDAG] Add preliminary plumbing for `samesign` flag

### PR DESCRIPTION
**Pull request by antoniofrighetto**
Extend recently-added poison-generating IR flag to codegen as well.

I think there are some other places in TargetLowering (e.g., `LegalizeSetCCCondCode`) where the flag should be cleared, welcoming directions (also maybe besides `SDNodeFlags`, it can be added to `MIFlag` directly here).

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
Missed operator==. Should InstrEmitter propagate it?

**Comment by antoniofrighetto** [comment]
> Missed operator==. Should InstrEmitter propagate it?

Added, thanks! Extended it to MIRParser as well.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 1]
```diff
@@ -3640,6 +3640,10 @@ void SelectionDAGBuilder::visitICmp(const ICmpInst &I) {
   SDValue Op2 = getValue(I.getOperand(1));
   ISD::CondCode Opcode = getICmpCondCode(predicate);
 
+  SDNodeFlags Flags;
+  Flags.setSameSign(I.hasSameSign());
+  SelectionDAG::FlagInserter FlagsInserter(DAG, Flags);
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:3645]: I think this needs to be below the possible `getPtrExtOrTrunc` below.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 2]
```diff
@@ -1716,8 +1717,13 @@ bool TargetLowering::SimplifyDemandedBits(
       // if we don't care about FP signed-zero. The use of SETLT with FP means
       // that we don't care about NaNs.
       if (CC == ISD::SETLT && Op1.getValueType().isInteger() &&
-          (isNullConstant(Op1) || ISD::isBuildVectorAllZeros(Op1.getNode())))
+          (isNullConstant(Op1) || ISD::isBuildVectorAllZeros(Op1.getNode()))) {
+        if (Flags.hasSameSign()) {
+          Flags.setSameSign(false);
+          Op->setFlags(Flags);
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:1723]: This doesn't seem necessary for the plumbing and its not clear its correct to propagate the rest of the flags.

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:1723]: Don't we want to stop propagating the flag in `SimplifyDemandedBits`?

File: llvm/lib/CodeGen/MIRPrinter.cpp [hunk 3]
```diff
@@ -837,6 +837,8 @@ void MIPrinter::print(const MachineInstr &MI) {
     OS << "disjoint ";
   if (MI.getFlag(MachineInstr::NoUSWrap))
     OS << "nusw ";
+  if (MI.getFlag(MachineInstr::SameSign))
```
Comment (arsenm) [inline llvm/lib/CodeGen/MIRPrinter.cpp:840]: Should get mir print / parse test

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 2 above)
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:1723]: Needs test and is beyond infrastructure work

Comment (goldsteinn) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:1723]: Ahh, I misunderstood earlier.

I don't think we need to drop `Op` flags here. We aren't simplifying `Op0` or `Op1` here, just replacing `Op` with `Op0`.

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp]: That makes sense, dropped it.

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp]: Dropped it.

File: llvm/lib/CodeGen/MIRPrinter.cpp (hunk 3 above)
Comment (antoniofrighetto) [reply llvm/lib/CodeGen/MIRPrinter.cpp:840]: Testing this via GISel.

Comment (arsenm) [reply llvm/lib/CodeGen/MIRPrinter.cpp:840]: This test doesn't cover the parsing, see test/CodeGen/MIR/X86/fastmath.mir as an example

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/MIRPrinter.cpp:840]: Added, thanks, think it should be on track now.

File: llvm/test/CodeGen/MIR/X86/icmp-samesign.mir [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:gr32 = COPY $esi
+    ; CHECK-NEXT: [[COPY1:%[0-9]+]]:gr32 = COPY $edi
+    ; CHECK-NEXT: samesign CMP32rr killed [[COPY1]], killed [[COPY]], implicit-def $eflags
+    ; CHECK-NEXT: $al = SETCCr 2, implicit killed $eflags
+    ; CHECK-NEXT: RET 0, $al
+    %1 = COPY $esi
+    %0 = COPY $edi
+    samesign CMP32rr killed %0, killed %1, implicit-def $eflags
+    $al = SETCCr 2, implicit killed $eflags
```
Comment (arsenm) [inline llvm/test/CodeGen/MIR/X86/icmp-samesign.mir:36]: Would be more sensible to have this on a G_MIR instruction as a sample

**Review by arsenm — APPROVED** [review]
I feel like I just reviewed the same thing in #113090

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 5]
```diff
@@ -3652,6 +3652,10 @@ void SelectionDAGBuilder::visitICmp(const ICmpInst &I) {
     Op2 = DAG.getPtrExtOrTrunc(Op2, getCurSDLoc(), MemVT);
   }
 
+  SDNodeFlags Flags;
+  Flags.setSameSign(I.hasSameSign());
+  SelectionDAG::FlagInserter FlagsInserter(DAG, Flags);
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:3657]: Will need update after #114061

File: llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-icmp-samesign.ll [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  ; CHECK-NEXT:   liveins: $w0, $w1
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:_(s32) = COPY $w0
+  ; CHECK-NEXT:   [[COPY1:%[0-9]+]]:_(s32) = COPY $w1
+  ; CHECK-NEXT:   %2:_(s1) = samesign G_ICMP intpred(ult), [[COPY]](s32), [[COPY1]]
+  ; CHECK-NEXT:   [[ZEXT:%[0-9]+]]:_(s8) = G_ZEXT %2(s1)
+  ; CHECK-NEXT:   [[ANYEXT:%[0-9]+]]:_(s32) = G_ANYEXT [[ZEXT]](s8)
+  ; CHECK-NEXT:   $w0 = COPY [[ANYEXT]](s32)
+  ; CHECK-NEXT:   RET_ReallyLR implicit $w0
+    %res = icmp samesign ult i32 %a, %b
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-icmp-samesign.ll:16]: There is no test without `samesign`.

File: llvm/test/CodeGen/MIR/X86/icmp-samesign.mir [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    ; CHECK: liveins: $edi, $esi
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: %tmp0:_(s32) = COPY $edi
+    ; CHECK-NEXT: %tmp1:_(s32) = COPY $esi
+    ; CHECK-NEXT: %tmp4:_(s8) = samesign G_ICMP intpred(ult), %tmp0(s32), %tmp1
+    ; CHECK-NEXT: $al = COPY %tmp4(s8)
+    ; CHECK-NEXT: RET 0, implicit $al
+    %tmp0:_(s32) = COPY $edi
+    %tmp1:_(s32) = COPY $esi
+    %tmp4:_(s8) = samesign G_ICMP intpred(ult), %tmp0(s32), %tmp1
```
Comment (tschuett) [inline llvm/test/CodeGen/MIR/X86/icmp-samesign.mir:36]: Again no test without `samesign` .

**Comment by tschuett** [comment]
If everything works well this time, you can rebase and limit the changes to SDAG.

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 8]
```diff
@@ -340,18 +340,18 @@ bool IRTranslator::translateCompare(const User &U,
   Register Op1 = getOrCreateVReg(*U.getOperand(1));
   Register Res = getOrCreateVReg(U);
   CmpInst::Predicate Pred = CI->getPredicate();
+  uint32_t Flags = 0;
+  if (CI)
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:344]: `CI` is never `nullptr`, but it will disappear during rebase anyway.

File: llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+; RUN: llc -march=aarch64 -stop-after=finalize-isel -simplify-mir -o - %s | FileCheck %s
+
+define i1 @icmp_samesign(i32 %a, i32 %b) {
+  ; CHECK-LABEL: name: icmp_samesign
+  ; CHECK: bb.0 (%ir-block.0):
+  ; CHECK-NEXT:   liveins: $w0, $w1
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:gpr32 = COPY $w1
+  ; CHECK-NEXT:   [[COPY1:%[0-9]+]]:gpr32 = COPY $w0
+  ; CHECK-NEXT:   [[SUBSWrr:%[0-9]+]]:gpr32 = SUBSWrr [[COPY1]], [[COPY]], implicit-def $nzcv
```
Comment (antoniofrighetto) [inline llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll:11]: @arsenm Was trying to test DAG preserving of the flag (to test InstrEmitter), but it seems like the flag gets lost during legalization of the DAG (which boils down to a custom LowerSETCC for AArch64/X86). Shouldn't we want to preserve the flag there as well, or is it beyond patch scope?

Comment (arsenm) [reply llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll:11]: The value of preserving the flag is diminishing the further back you go. I doubt anything after selection will have any use of this

File: llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h [hunk 10]
```diff
@@ -1267,7 +1267,7 @@ class MachineIRBuilder {
   /// \return a MachineInstrBuilder for the newly created instruction.
   MachineInstrBuilder buildICmp(CmpInst::Predicate Pred, const DstOp &Res,
                                 const SrcOp &Op0, const SrcOp &Op1,
-                                std::optional<unsigned> Flgs = std::nullopt);
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1270]: Unrelated typo fix

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp (hunk 5 above)
Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:3657]: This was committed so can do this

File: llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  ; CHECK-LABEL: name: icmp_samesign_vec
+  ; CHECK: bb.0 (%ir-block.0):
+  ; CHECK-NEXT:   liveins: $d0, $d1
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:fpr64 = COPY $d1
+  ; CHECK-NEXT:   [[COPY1:%[0-9]+]]:fpr64 = COPY $d0
+  ; CHECK-NEXT:   [[CMHIv2i32_:%[0-9]+]]:fpr64 = CMHIv2i32 [[COPY]], [[COPY1]]
+  ; CHECK-NEXT:   $d0 = COPY [[CMHIv2i32_]]
+  ; CHECK-NEXT:   RET_ReallyLR implicit $d0
+  %res = icmp samesign ult <2 x i32> %a, %b
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll:29]: Could you please add a test without samesign?

File: llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h (hunk 10 above)
Comment (antoniofrighetto) [reply llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1270]: I took the minor opportunity to fix the typo within this patch, will undo if you have a strong concern on this.

Comment (tschuett) [reply llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1270]: That was probably me. I vote for fix.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp (hunk 5 above)
Comment (antoniofrighetto) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:3657]: We are still free to use FlagInserter, independently of the changes in #114061, this should be fine.

File: llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll (hunk 9 above)
Comment (antoniofrighetto) [reply llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll:11]: Shouldn't we still preserve the flag immediately before doing instruction selection (before the very last iteration of DAG combining)?

Comment (arsenm) [reply llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll:11]: Don't really care, it's easy enough to handle. I would agree preserving the flag in other contexts is beyond the scope of this patch

Comment (antoniofrighetto) [reply llvm/test/CodeGen/AArch64/dag-preserve-icmp-samesign.ll:11]: Okay, I'm dropping the test then, and merging this, if you are still fine with everything else.

**Comment by gulfemsavrun** [comment]
This PR caused an assertion error:
```
[1321/5119](49) Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o
FAILED: lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o 
/b/s/w/ir/x/w/llvm_build/./bin/clang++ --sysroot=/b/s/w/ir/x/w/cipd/linux -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/lib/Analysis -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/include -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/include -isystem /b/s/w/ir/x/w/zlib_install_target/include -isystem /b/s/w/ir/x/w/zstd_install/include -stdlib=libc++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -flto=full -ffat-lto-objects -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins=../../../../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -O3 -DNDEBUG -std=c++17 -fvisibility=default  -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -MD -MT lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -MF lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o.d -o lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -c /b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis/BranchProbabilityInfo.cpp
clang++: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:16157: SDValue (anonymous namespace)::DAGCombiner::visitFREEZE(SDNode *): Assertion `DAG.isGuaranteedNotToBeUndefOrPoison(R, false) && "Can't create node that may be undef/poison!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /b/s/w/ir/x/w/llvm_build/./bin/clang++ --sysroot=/b/s/w/ir/x/w/cipd/linux -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/lib/Analysis -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/include -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/include -isystem /b/s/w/ir/x/w/zlib_install_target/include -isystem /b/s/w/ir/x/w/zstd_install/include -stdlib=libc++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -flto=full -ffat-lto-objects -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins=../../../../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -O3 -DNDEBUG -std=c++17 -fvisibility=default -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -MD -MT lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -MF lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o.d -o lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -c /b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis/BranchProbabilityInfo.cpp
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '/b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis/BranchProbabilityInfo.cpp'.
4.	Running pass 'AArch64 Instruction Selection' on function '@_ZN4llvm21BranchProbabilityInfo19calcMetadataWeightsEPKNS_10BasicBlockE'
#0 0x0000aaaacdc240c0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/b/s/w/ir/x/w/llvm_build/./bin/clang+++0x42040c0)
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
```
https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8732522040154573889/+/u/clang/build/stdout

**Comment by antoniofrighetto** [comment]
> This PR caused an assertion error:
> 
> ```
> [1321/5119](49) Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o
> FAILED: lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o 
> /b/s/w/ir/x/w/llvm_build/./bin/clang++ --sysroot=/b/s/w/ir/x/w/cipd/linux -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/lib/Analysis -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/include -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/include -isystem /b/s/w/ir/x/w/zlib_install_target/include -isystem /b/s/w/ir/x/w/zstd_install/include -stdlib=libc++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -flto=full -ffat-lto-objects -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins=../../../../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -O3 -DNDEBUG -std=c++17 -fvisibility=default  -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -MD -MT lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -MF lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o.d -o lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -c /b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis/BranchProbabilityInfo.cpp
> clang++: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:16157: SDValue (anonymous namespace)::DAGCombiner::visitFREEZE(SDNode *): Assertion `DAG.isGuaranteedNotToBeUndefOrPoison(R, false) && "Can't create node that may be undef/poison!"' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
> Stack dump:
> 0.	Program arguments: /b/s/w/ir/x/w/llvm_build/./bin/clang++ --sysroot=/b/s/w/ir/x/w/cipd/linux -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/lib/Analysis -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis -I/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/include -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/include -isystem /b/s/w/ir/x/w/zlib_install_target/include -isystem /b/s/w/ir/x/w/zstd_install/include -stdlib=libc++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -flto=full -ffat-lto-objects -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins=../../../../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -O3 -DNDEBUG -std=c++17 -fvisibility=default -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -MD -MT lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -MF lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o.d -o lib/Analysis/CMakeFiles/LLVMAnalysis.dir/BranchProbabilityInfo.cpp.o -c /b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis/BranchProbabilityInfo.cpp
> 1.	<eof> parser at end of file
> 2.	Code generation
> 3.	Running pass 'Function Pass Manager' on module '/b/s/w/ir/x/w/llvm-llvm-project/llvm/lib/Analysis/BranchProbabilityInfo.cpp'.
> 4.	Running pass 'AArch64 Instruction Selection' on function '@_ZN4llvm21BranchProbabilityInfo19calcMetadataWeightsEPKNS_10BasicBlockE'
> #0 0x0000aaaacdc240c0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/b/s/w/ir/x/w/llvm_build/./bin/clang+++0x42040c0)
> clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
> ```
> 
> https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8732522040154573889/+/u/clang/build/stdout

Patch attempting to fix the issue: https://github.com/llvm/llvm-project/pull/114582.
